File: go-broker/internal/physics/integrator.go
package physics

import (
	"math"

	pb "driftpursuit/broker/internal/proto/pb"
)

// Vec3 is a lightweight vector helper used by the physics utilities.
type Vec3 struct {
	X float64
	Y float64
	Z float64
}

// FromProtoVec3 converts a protobuf vector into the helper representation.
func FromProtoVec3(v *pb.Vector3) Vec3 {
	//1.- Protect against nil inputs to keep call sites concise.
	if v == nil {
		return Vec3{}
	}
	return Vec3{X: v.X, Y: v.Y, Z: v.Z}
}

// ToProtoVec3 copies the helper representation back into a protobuf vector.
func ToProtoVec3(v Vec3, dst *pb.Vector3) *pb.Vector3 {
	//1.- Allocate the destination if required for easier callers.
	if dst == nil {
		dst = &pb.Vector3{}
	}
	//2.- Copy each component for the mutation based workflow.
	dst.X = v.X
	dst.Y = v.Y
	dst.Z = v.Z
	return dst
}

// wrapAngleDeg normalizes an angle to the [-180, 180) range.
func wrapAngleDeg(angle float64) float64 {
	//1.- Use math.Mod to keep values bounded across many integration steps.
	wrapped := math.Mod(angle+180.0, 360.0)
	if wrapped < 0 {
		wrapped += 360.0
	}
	return wrapped - 180.0
}

// integrateLinear applies velocity over the timestep to update the position.
func integrateLinear(position *pb.Vector3, velocity *pb.Vector3, step float64) {
	//1.- Skip integration when inputs are missing or invalid.
	if position == nil || velocity == nil || step <= 0 {
		return
	}
	//2.- Advance each axis using the standard Euler integration.
	position.X += velocity.X * step
	position.Y += velocity.Y * step
	position.Z += velocity.Z * step
}

// integrateAngular applies angular velocity to the Euler orientation.
func integrateAngular(orientation *pb.Orientation, angularVelocity *pb.Vector3, step float64) {
	//1.- Require valid orientation data before attempting integration.
	if orientation == nil || angularVelocity == nil || step <= 0 {
		return
	}
	//2.- Update each Euler component in degrees per second then wrap.
	orientation.YawDeg = wrapAngleDeg(orientation.YawDeg + angularVelocity.Y*step)
	orientation.PitchDeg = wrapAngleDeg(orientation.PitchDeg + angularVelocity.X*step)
	orientation.RollDeg = wrapAngleDeg(orientation.RollDeg + angularVelocity.Z*step)
}

// IntegrateVehicle advances both linear and angular state for the vehicle.
func IntegrateVehicle(state *pb.VehicleState, step float64) {
	//1.- Guard against nil or invalid timesteps for robustness.
	if state == nil || step <= 0 {
		return
	}
	//2.- Integrate translation if both position and velocity are present.
	integrateLinear(state.Position, state.Velocity, step)
	//3.- Integrate rotation using the angular velocity channel.
	integrateAngular(state.Orientation, state.AngularVelocity, step)
}

File: typescript-client/src/physics/integrator.ts
export interface Vec3 {
  x: number;
  y: number;
  z: number;
}

export interface OrientationDeg {
  yawDeg: number;
  pitchDeg: number;
  rollDeg: number;
}

export interface VehicleStateLike {
  position?: Vec3;
  velocity?: Vec3;
  orientation?: OrientationDeg;
  angularVelocity?: Vec3;
  flightAssistEnabled?: boolean;
}

// wrapAngleDeg normalizes an angle into the [-180, 180) interval.
export function wrapAngleDeg(angle: number): number {
  //1.- Use modulo arithmetic to prevent unbounded growth over long runs.
  const wrapped = ((angle + 180) % 360 + 360) % 360;
  return wrapped - 180;
}

// integrateLinear advances a position using simple Euler integration.
export function integrateLinear(position: Vec3 | undefined, velocity: Vec3 | undefined, step: number): void {
  //1.- Guard against invalid inputs so callers can pass partial state objects.
  if (!position || !velocity || !(step > 0)) {
    return;
  }
  //2.- Apply the displacement derived from velocity * dt on each axis.
  position.x += velocity.x * step;
  position.y += velocity.y * step;
  position.z += velocity.z * step;
}

// integrateAngular updates Euler angles from angular velocity in degrees/s.
export function integrateAngular(orientation: OrientationDeg | undefined, angularVelocity: Vec3 | undefined, step: number): void {
  //1.- Skip when the vehicle lacks rotation data or the step is degenerate.
  if (!orientation || !angularVelocity || !(step > 0)) {
    return;
  }
  //2.- Add the integrated deltas and wrap to keep the values bounded.
  orientation.yawDeg = wrapAngleDeg(orientation.yawDeg + angularVelocity.y * step);
  orientation.pitchDeg = wrapAngleDeg(orientation.pitchDeg + angularVelocity.x * step);
  orientation.rollDeg = wrapAngleDeg(orientation.rollDeg + angularVelocity.z * step);
}

// integrateVehicle mutates the provided state with both linear and angular updates.
export function integrateVehicle(state: VehicleStateLike | undefined, step: number): void {
  //1.- Support defensive callers by no-oping on invalid state or timestep.
  if (!state || !(step > 0)) {
    return;
  }
  //2.- Apply both translation and rotation integration in place.
  integrateLinear(state.position, state.velocity, step);
  integrateAngular(state.orientation, state.angularVelocity, step);
}

export class GuidanceSpline {
  private readonly nodes: Vec3[];

  constructor(nodes: Vec3[]) {
    //1.- Require at least two nodes so a tangent can be computed.
    if (nodes.length < 2) {
      throw new Error("GuidanceSpline requires at least two nodes");
    }
    //2.- Copy the input to prevent external mutation after construction.
    this.nodes = nodes.map((node) => ({ ...node }));
  }

  private tangentFor(position: Vec3): Vec3 | undefined {
    //1.- Track the closest segment to the provided position.
    let bestDistance = Number.POSITIVE_INFINITY;
    let bestTangent: Vec3 | undefined;
    for (let i = 0; i < this.nodes.length - 1; i += 1) {
      const a = this.nodes[i];
      const b = this.nodes[i + 1];
      const ab = { x: b.x - a.x, y: b.y - a.y, z: b.z - a.z };
      const abLenSquared = ab.x * ab.x + ab.y * ab.y + ab.z * ab.z;
      if (abLenSquared === 0) {
        continue;
      }
      const ap = { x: position.x - a.x, y: position.y - a.y, z: position.z - a.z };
      let t = (ap.x * ab.x + ap.y * ab.y + ap.z * ab.z) / abLenSquared;
      if (t < 0) {
        t = 0;
      } else if (t > 1) {
        t = 1;
      }
      const closest = { x: a.x + ab.x * t, y: a.y + ab.y * t, z: a.z + ab.z * t };
      const dx = position.x - closest.x;
      const dy = position.y - closest.y;
      const dz = position.z - closest.z;
      const distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
      if (distance < bestDistance) {
        bestDistance = distance;
        const length = Math.sqrt(abLenSquared);
        bestTangent = { x: ab.x / length, y: ab.y / length, z: ab.z / length };
      }
    }
    return bestTangent;
  }

  align(state: VehicleStateLike): void {
    //1.- Only operate when both the spline and vehicle position are available.
    if (!state.position) {
      return;
    }
    const tangent = this.tangentFor(state.position);
    if (!tangent) {
      return;
    }
    //2.- Compute yaw and pitch angles from the tangent direction.
    const horizontal = Math.sqrt(tangent.x * tangent.x + tangent.z * tangent.z);
    const yaw = horizontal === 0 ? 0 : (Math.atan2(tangent.x, tangent.z) * 180) / Math.PI;
    const pitch = (Math.atan2(tangent.y, horizontal) * 180) / Math.PI;
    //3.- Allocate orientation if necessary and zero angular velocity for stability.
    if (!state.orientation) {
      state.orientation = { yawDeg: 0, pitchDeg: 0, rollDeg: 0 };
    }
    state.orientation.yawDeg = wrapAngleDeg(yaw);
    state.orientation.pitchDeg = wrapAngleDeg(pitch);
    state.orientation.rollDeg = 0;
    if (state.angularVelocity) {
      state.angularVelocity.x = 0;
      state.angularVelocity.y = 0;
      state.angularVelocity.z = 0;
    }
  }
}

// applyAssistAlignment conditionally aligns the craft when assist is active.
export function applyAssistAlignment(state: VehicleStateLike | undefined, spline: GuidanceSpline | undefined): void {
  //1.- Skip unless assist mode is toggled and the spline is available.
  if (!state || !state.flightAssistEnabled || !spline) {
    return;
  }
  //2.- Delegate to the spline alignment helper.
  spline.align(state);
}
